File: DS/SORTING/MergeSort.py
a = [9,3,2,7,6,7,7,11]
n = len(a)

def mergeSort(a, start,end):
    if start<end:
        mid = int((start+end)/2)
        mergeSort(a,start,mid)
        mergeSort(a,mid+1,end)
        merge(a,start,mid,end)

def merge(a, start, mid, end):
    i = start
    j = mid+1
    arr = []
    while i<=mid and j<=end:
        if a[i]<=a[j]:
            arr.append(a[i])
            i +=1
        else:
            arr.append(a[j])
            j +=1

    if i<=mid:
        while i<=mid:
            arr.append(a[i])
            i +=1
    if j<=end:
        while j<=mid:
            arr.append(a[j])
            j +=1

    k = start
    for i in arr:
        a[k] = i
        k +=1

mergeSort(a,0,n-1)
print(a)

File: DS/SORTING/MergeSort.java
 import java.util.List;
 import java.util.ArrayList;
 class MergeSort{

     static void sort(int[]a, int start, int end){
         if(start<end){
             int mid = (start+ end)/2;
             sort(a,start,mid);
             sort(a,mid+1,end);
             merge(a,start,mid,end);
         }
     }

    static void merge(int[] a, int start, int mid, int end){
        int i = start, j = mid+1;
        List<Integer> ls = new ArrayList<>();
        while(i<=mid && j<=end){
            if(a[i]<=a[j]){
                ls.add(a[i]);
                i++;
            }else{
                ls.add(a[j]);
                j++;
            }
        }
        if(i<=mid){
            while(i<=mid){
                ls.add(a[i]);
                i++;
            }
        }
       if(j<=end){
            while(j<=mid){
                ls.add(a[j]);
                j++;
            }
        }
        for(Integer e:ls){
            a[start] = e;
            start++;
        }
    }

     public static void main(String args[]){
        int a[] = {7,3,4,6,7,1,2,3};
        sort(a,0,a.length-1);
        System.out.println(java.util.Arrays.toString(a));
     }


 }